src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
I think getting a nice API is a good idea. It's easy to provide overloads that transform primitive and common F# CLR values into the object model we both linked to, but not as easy when they are complex without letting the author of the type extend the type with an extra serialisation method.

```
module Value =
  let ofCLRValue (data : ^a) =
   // ...
// and 
let addData (m : Message) (key : String) (data : ^a) =
 Lens.setLens (Message.DataLens key) (Value.ofCLRValue data)
```

Could that work? I'm thinking we can provide the (bind, apply) applicatives for a new Value monad, like Chrion does, which lets you compose the values properly; thereby both defining the wire-format as well as letting MessageTemplates use all of the structure.

As for the structure; MessageTemplates is using a subset of the Value type that I linked to. How about we use the superset of mine in this library? That will allow supporting metrics easier, like is Logary's prerogative.